src/Education/XamlGenerator/CustomTextMarkupParser.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarLint.VisualStudio.Core;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+
+/* Notes:
+ *   
+ *   The SonarQube and SonarCloud suppport a special embedded syntax to indicate
+ *   cross-references between links
+ *   
+ *   e.g. https://sonarcloud.io/organizations/sonarsource/rules?languages=cs&open=csharpsquid%3AS4002&q=S4002
+ * 
+ *   The cross-references are embedded in the rule description text as follows:
+ *      {rule:[repoKey]:[ruleKey]}
+ * 
+ *   This class parses simple text strings looking for the embedded syntax and
+ *   calls the appropriate callback so the caller can render the appropriate output.
+ */
+
+namespace SonarLint.VisualStudio.Education.XamlGenerator
+{
+    /// <summary>
+    /// Processes special internal markup embedded in normal text
+    /// </summary>
+    internal static class CustomTextMarkupParser
```

Comment:
We'll probably need to add an interface for this to make it easier to test the classes that use it.
However, that can be done in a later PR.